merkle/smt/hstar3.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package smt contains the implementation of the sparse Merkle tree logic.
+package smt
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/google/trillian/storage/tree"
+)
+
+// NodeStorage reads and writes sparse Merkle tree node hashes.
+type NodeStorage interface {
+	// Get returns the hash of the given node.
+	Get(id tree.NodeID2) ([]byte, error)
+	// Set sets the hash of the given node.
+	Set(id tree.NodeID2, hash []byte)
+}
+
+// HashChildrenFn computes a node hash based on its child nodes' hashes.
+type HashChildrenFn func(left, right []byte) []byte
```

Comment:
Not necessarily. It depends on how the thing overarching `HStar3` is set up. If it's parallel then `HashChildren` must be thread-safe, but if not then it doesn't have to. If I added such a comment, it would be 1 level above in the sparse Merkle tree writer.